backend/src/main/java/ca/bc/gov/hlth/mohums/controller/UsersController.java

```diff
@@ -103,16 +103,31 @@ public ResponseEntity<List<Object>> getUsers(
                 Long lastLogBeforeEpoch = LocalDate.parse(lastLogBefore.get()).atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();
                 lastLogBeforeCriteria = Optional.of(" HAVING MAX(event_time) < " + lastLogBeforeEpoch);
             }
+            Optional<String> clientAndRolesJoins = Optional.empty();
+            Optional<String> clientAndRolesCriteria = Optional.empty();
+            if(clientName.isPresent()){
+                clientAndRolesJoins = Optional.of(
+                        " JOIN KEYCLOAK.USER_ROLE_MAPPING urm ON urm.USER_ID = ee.USER_ID"
+                                + " JOIN KEYCLOAK.KEYCLOAK_ROLE kr ON kr.ID = urm.ROLE_ID "
+                                + " JOIN KEYCLOAK.CLIENT c ON c.ID = kr.CLIENT"
+                );
+                clientAndRolesCriteria = Optional.of(
+                        " AND ee.CLIENT_ID = " + "'" + clientName.get() + "'"
+                        + " AND kr.NAME IN " + listOfRoles(selectedRoles, clientId.get())
+                );
+            }
+            String doNotInclude = "";
 
             String sql
-                    = "SELECT user_id, MAX(event_time) AS last_login"
-                    + "  FROM keycloak.event_entity"
-                    + " WHERE type='LOGIN'"
-                    + "   AND user_id IS NOT NULL" + customDateCriteria
-                    + " GROUP BY user_id";
-            if(lastLogBeforeCriteria.isPresent()){
-                sql += lastLogBeforeCriteria.get();
-            }
+                    = "SELECT ee.user_id, MAX(ee.event_time) AS last_login"
+                    + " FROM keycloak.event_entity ee"
+                    + clientAndRolesJoins.orElse(doNotInclude)
+                    + " WHERE ee.type='LOGIN'"
+                    + " AND ee.user_id IS NOT NULL" + customDateCriteria
+                    + clientAndRolesCriteria.orElse(doNotInclude)
+                    + " GROUP BY ee.user_id"
+                    + lastLogBeforeCriteria.orElse(doNotInclude);
```

Comment:
I agree with what you said earlier and this whole implementation could use some refactoring. That said I'm glad you didn't as this is easier to review for correctness.